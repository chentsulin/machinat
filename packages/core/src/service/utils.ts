import invariant from 'invariant';
import {
  MACHINAT_SERVICE_CONTAINER,
  MACHINAT_SERVICE_PROVIDER,
  MACHINAT_SERVICE_INTERFACE,
} from '../symbol';
import ServiceScope from './scope';
import ServiceMaker from './maker';
import ProvisionMap from './provisionMap';
import type {
  Interfaceable,
  ServiceRequirement,
  ServiceContainer,
  ServiceProvider,
} from './types';

export const isServiceContainer = <T>(
  target: T | ServiceContainer<T>
): target is ServiceContainer<T> =>
  typeof target === 'function' &&
  '$$typeof' in target &&
  target.$$typeof === MACHINAT_SERVICE_CONTAINER;

export const isServiceProvider = (
  target: any
): target is ServiceProvider<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  target.$$typeof === MACHINAT_SERVICE_PROVIDER;

export const maybeInjectContainer = <T>(
  scope: ServiceScope,
  maybeContainer: T | ServiceContainer<T>
): T =>
  isServiceContainer(maybeContainer)
    ? scope.injectContainer(maybeContainer)
    : maybeContainer;

export const isInterfaceable = (target: any): target is Interfaceable<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  (target.$$typeof === MACHINAT_SERVICE_INTERFACE ||
    target.$$typeof === MACHINAT_SERVICE_PROVIDER);

export const createEmptyScope = (platform?: string): ServiceScope =>
  new ServiceScope(new ServiceMaker(new ProvisionMap()), new Map());

/** @internal */
export const polishServiceRequirement = <T>(
  dep: Interfaceable<T> | ServiceRequirement<T>
): ServiceRequirement<T> => {
  if (isInterfaceable(dep)) {
    return { require: dep, optional: false };
  }

  invariant(
    'require' in dep &&
      (dep.require.$$typeof === MACHINAT_SERVICE_INTERFACE ||
        dep.require.$$typeof === MACHINAT_SERVICE_PROVIDER),
    `${(dep as any).name || String(dep)} is not a valid interface`
  );

  return dep;
};
