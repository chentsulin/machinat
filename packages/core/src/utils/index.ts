export { default as compose } from './compose';
export { default as filterSymbolKeys } from './filterSymbolKeys';
export { default as formatNode } from './formatNode';
export * from './isX';
export { default as mixin } from './mixin';
