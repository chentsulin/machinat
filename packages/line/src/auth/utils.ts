/** @internal */ /** */
/* eslint-disable import/prefer-default-export  */
import LineUser from '../user';
import LineChat from '../channel';
import { LIFFAuthData, AuthorizerRefinement } from './types';

export const refinementFromLIFFAuthData = (
  providerId: string,
  botChannelId: string,
  data: LIFFAuthData
): null | AuthorizerRefinement => {
  const { contextType, userId, utouId, groupId, roomId, botChannel } = data;

  let channel: null | LineChat;

  if (botChannel) {
    if (botChannel !== botChannelId || contextType !== 'utou') {
      return null;
    }

    channel = new LineChat(botChannelId, 'utob', userId);
  } else {
    channel =
      contextType === 'utou'
        ? new LineChat(botChannelId, 'utou', utouId as string)
        : contextType === 'room'
        ? new LineChat(botChannelId, 'room', roomId as string)
        : contextType === 'group'
        ? new LineChat(botChannelId, 'group', groupId as string)
        : null;
  }

  return {
    user: new LineUser(providerId, userId),
    channel,
  };
};
