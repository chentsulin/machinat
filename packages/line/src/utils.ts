import { annotateNativeComponent } from '@machinat/core/renderer';
import type { LineSegmentValue, LineMessageSegmentValue } from './types';
import { LINE, CHANNEL_REQUEST_GETTER } from './constant';

export const annotateLineComponent = annotateNativeComponent(LINE);

export const isMessageValue = (
  value: string | LineSegmentValue
): value is string | LineMessageSegmentValue =>
  typeof value === 'string' ||
  (typeof value === 'object' &&
    !Object.prototype.hasOwnProperty.call(value, CHANNEL_REQUEST_GETTER));
