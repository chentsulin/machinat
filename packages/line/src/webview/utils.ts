/** @internal */ /** */
/* eslint-disable import/prefer-default-export  */
import LineUser from '../user';
import LineChat from '../channel';
import { LineAuthContext, AuthorizerRefinement } from './types';

export const refineAuthContext = ({
  contextType,
  userId,
  groupId,
  roomId,
  channelId,
  providerId,
}: LineAuthContext): null | AuthorizerRefinement => ({
  user: new LineUser(providerId, userId),
  channel:
    contextType === 'group'
      ? new LineChat(channelId, 'group', groupId as string)
      : contextType === 'room'
      ? new LineChat(channelId, 'room', roomId as string)
      : new LineChat(channelId, 'user', userId),
});
