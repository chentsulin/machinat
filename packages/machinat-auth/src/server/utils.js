// @flow
import type { IncomingMessage, ServerResponse } from 'http';
import { parse as parseCookie, serialize as serializeCookie } from 'cookie';

export const respondRedirect = (res: ServerResponse, url: string) => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (
  req: IncomingMessage
): null | {| [string]: string |} => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: Object
) => {
  res.setHeader('Set-Cookie', serializeCookie(key, value, options));
};

export const checkDomainScope = (subdomain: string, domain: string) =>
  subdomain.slice(-domain.length) === domain &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const checkPathScope = (subpath: string, path: string) =>
  subpath.slice(0, path.length) === path &&
  (subpath.length === path.length ||
    path[path.length - 1] === '/' ||
    subpath[subpath.length - path.length] === '/');
