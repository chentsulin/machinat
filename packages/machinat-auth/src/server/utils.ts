import type { ServerResponse } from 'http';
import {
  parse as parseCookie,
  serialize as serializeCookie,
  CookieSerializeOptions,
} from 'cookie';
import type { WithHeaders } from '../types';

export const respondRedirect = (res: ServerResponse, url: string): void => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (
  req: WithHeaders
): null | {
  [key: string]: string;
} => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

const SET_COOKIE = 'Set-Cookie';

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: CookieSerializeOptions
): void => {
  const cookieDesc = serializeCookie(key, value, options);
  const cookiesAlreadySet = res.getHeader(SET_COOKIE);

  res.setHeader(
    SET_COOKIE,
    Array.isArray(cookiesAlreadySet)
      ? [...cookiesAlreadySet, cookieDesc]
      : typeof cookiesAlreadySet === 'string'
      ? [cookiesAlreadySet, cookieDesc]
      : cookieDesc
  );
};

export const isSubdomain = (domain: string, subdomain: string): boolean =>
  subdomain.slice(-domain.length) === domain &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const isSubpath = (path: string, subpath: string): boolean =>
  subpath.slice(0, path.length) === path &&
  (subpath.length === path.length ||
    path[path.length - 1] === '/' ||
    subpath[path.length] === '/');
