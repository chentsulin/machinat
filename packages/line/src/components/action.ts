import { partSegment } from '@machinat/core/renderer';
import { PartSegment } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import type { LineComponent } from '../types';

/**
 * @category Props
 */
type PostbackActionProps = {
  /** Label for the action */
  label: string;
  /**
   * String returned via webhook in the `postback.data` property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /**
   * Text displayed in the chat as a message sent by the user when the action is
   * performed. Required for quick reply buttons. Optional for the other message
   * types. Max character limit: 300.
   */
  displayText?: string;
  /** Alias of `displayText` */
  text?: string;
};

/** @internal */
const __PostbackAction = function PostbackAction(node, path) {
  const { label, data, text, displayText } = node.props;
  return [
    partSegment(node, path, {
      type: 'postback',
      data,
      label,
      displayText: displayText || text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the specified string in the data property.
 * @category Component
 * @props {@link PostbackActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#postback-action).
 */
export const PostbackAction: LineComponent<
  PostbackActionProps,
  PartSegment<any>
> = annotateLineComponent(__PostbackAction);

/**
 * @category Props
 */
type MessageActionProps = {
  /** Label for the action */
  label: string;
  /** Text sent when the action is performed. Max character limit: 300 */
  text?: string;
};

/** @internal */
const __MessageAction = function MessageAction(node, path) {
  const { label, text } = node.props;
  return [
    partSegment(node, path, {
      type: 'message',
      label,
      text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the string in the
 * `text` property is sent as a message from the user.
 * @category Component
 * @props {@link MessageActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#message-action).
 */
export const MessageAction: LineComponent<
  MessageActionProps,
  PartSegment<any>
> = annotateLineComponent(__MessageAction);

/**
 * @category Props
 */
type UriActionProps = {
  /** Label for the action */
  label: string;
  /**
   * URI opened when the action is performed (Max character limit: 1000). The
   * available schemes are http, https, line, and tel.
   */
  uri: string;
};

/** @internal */
const __UriAction = function UriAction(node, path) {
  const { label, uri } = node.props;
  return [
    partSegment(node, path, {
      type: 'uri',
      label,
      uri,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the URI specified in
 * the `uri` property is opened.
 * @category Component
 * @props {@link UriActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#uri-action).
 */
export const UriAction: LineComponent<
  UriActionProps,
  PartSegment<any>
> = annotateLineComponent(__UriAction);

/** @internal */
const pad2 = (n) => (n < 10 ? `0${n}` : n);

/** @internal */
const fullDate = (d) =>
  `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;

/** @internal */
const fullHourMinute = (d) => `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;

/** @internal */
const dateToStringByMode = (mode, d) =>
  !(d instanceof Date)
    ? d
    : mode === 'datetime'
    ? `${fullDate(d)}T${fullHourMinute(d)}`
    : mode === 'date'
    ? fullDate(d)
    : fullHourMinute(d);

/**
 * @category Props
 */
type DateTimePickerActionProps = {
  /** Label for the action  */
  label: string;
  /**
   * String returned via webhook in the postback.data property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /** Action mode */
  mode: 'datetime' | 'date' | 'time';
  /** Initial value of date or time */
  initial?: string | Date;
  /**
   * Largest date or time value that can be selected. Must be greater than the
   * min value.
   */
  max?: string | Date;
  /**
   * Smallest date or time value that can be selected. Must be less than the max
   * value.
   */
  min?: string | Date;
};

/** @internal */
const __DateTimePickerAction = function DateTimePickerAction(node, path) {
  const { label, data, mode = 'datetime', initial, min, max } = node.props;

  return [
    partSegment(node, path, {
      type: 'datetimepicker',
      label,
      data,
      mode,
      initial: dateToStringByMode(mode, initial),
      max: dateToStringByMode(mode, max),
      min: dateToStringByMode(mode, min),
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the date and time selected by the user from the
 * date and time selection dialog. The datetime picker action does not support
 * time zones.
 * @category Component
 * @props {@link DateTimePickerActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#datetime-picker-action).
 */
export const DateTimePickerAction: LineComponent<
  DateTimePickerActionProps,
  PartSegment<any>
> = annotateLineComponent(__DateTimePickerAction);

/**
 * @category Props
 */
type CameraActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraAction = function CameraAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'camera',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera screen in LINE is opened.
 * @category Component
 * @props {@link CameraActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-action).
 */
export const CameraAction: LineComponent<
  CameraActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraAction);

/**
 * @category Props
 */
type CameraRollActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraRollAction = function CameraRollAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'cameraRoll',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera roll screen in LINE is
 * opened.
 * @category Component
 * @props {@link CameraRollActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const CameraRollAction: LineComponent<
  CameraRollActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraRollAction);

/**
 * @category Props
 */
type LocationActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __LocationAction = function LocationAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'location',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the location screen in LINE is opened.
 * @category Component
 * @props {@link LocationActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const LocationAction: LineComponent<
  LocationActionProps,
  PartSegment<any>
> = annotateLineComponent(__LocationAction);

export type Action =
  | typeof PostbackAction
  | typeof MessageAction
  | typeof UriAction
  | typeof DateTimePickerAction
  | typeof CameraAction
  | typeof CameraRollAction
  | typeof LocationAction;
