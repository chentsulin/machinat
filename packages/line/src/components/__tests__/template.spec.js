import Machinat from '@machinat/core';
import { isNativeType } from '@machinat/core/utils/isX';
import Renderer from '@machinat/core/renderer';
import {
  ButtonTemplate,
  ConfirmTemplate,
  CarouselItem,
  CarouselTemplate,
  ImageCarouselItem,
  ImageCarouselTemplate,
} from '../template';
import { UriAction } from '../action';

const renderer = new Renderer('line', () => null);

test.each([CarouselItem, ImageCarouselItem].map((C) => [C.name, C]))(
  '%s is valid native component',
  (_, Item) => {
    expect(typeof Item).toBe('function');

    expect(isNativeType(<Item />)).toBe(true);
    expect(Item.$$platform).toBe('line');
  }
);

test.each(
  [
    ButtonTemplate,
    ConfirmTemplate,
    CarouselTemplate,
    ImageCarouselTemplate,
  ].map((C) => [C.name, C])
)('%s is valid native unit component', (_, Template) => {
  expect(typeof Template).toBe('function');

  expect(isNativeType(<Template />)).toBe(true);
  expect(Template.$$platform).toBe('line');
});

test.each(
  [
    <ButtonTemplate
      altText="xxx"
      imageUrl="https://..."
      imageAspectRatio="square"
      imageSize="contain"
      imageBackgroundColor="#aaaaaa"
      title="HELLO"
      text="world"
      defaultAction={<UriAction uri="https://..." label="???" />}
      actions={[
        <UriAction uri="https://..." label="foo" />,
        <UriAction uri="https://..." label="bar" />,
      ]}
    />,

    <ConfirmTemplate
      altText="xxx"
      text="Take a pill"
      actions={[
        <UriAction uri="https://matrix.io/login" label="Blue pill" />,
        <UriAction uri="https://matrix.io/leave" label="Red pill" />,
      ]}
    />,

    <CarouselTemplate
      altText="xxx"
      imageAspectRatio="square"
      imageSize="contain"
    >
      <CarouselItem
        text="Burger"
        actions={[
          <UriAction uri="https://..." label="with fries" />,
          <UriAction uri="https://..." label="with salad" />,
        ]}
      />
      <CarouselItem
        imageUrl="https://..."
        imageBackgroundColor="#bbbbbb"
        title="Spaghetti"
        text="tamato sause"
        actions={[
          <UriAction uri="https://..." label="with soup" />,
          <UriAction uri="https://..." label="with salad" />,
        ]}
      />
    </CarouselTemplate>,

    <ImageCarouselTemplate altText="xxx">
      <ImageCarouselItem
        url="https://..."
        action={<UriAction uri="https://..." label="foo" />}
      />
      <ImageCarouselItem
        url="https://..."
        action={<UriAction uri="https://..." label="bar" />}
      />
    </ImageCarouselTemplate>,
  ].map((ele) => [ele.type.name, ele])
)('%s rendered match snapshot', async (_, templateElement) => {
  const promise = renderer.render(templateElement);
  await expect(promise).resolves.toEqual([
    {
      type: 'unit',
      node: templateElement,
      value: expect.any(Object),
      path: '$',
    },
  ]);

  const [{ value }] = await promise;
  expect(value).toMatchSnapshot();
});
