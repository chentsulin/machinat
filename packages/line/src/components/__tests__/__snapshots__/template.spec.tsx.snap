// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonTemplate render match snapshot 1`] = `
Object {
  "altText": "xxx",
  "template": Object {
    "actions": Array [
      Object {
        "label": "foo",
        "type": "uri",
        "uri": "https://...",
      },
      Object {
        "label": "bar",
        "type": "uri",
        "uri": "https://...",
      },
    ],
    "defaultAction": Object {
      "label": "???",
      "type": "uri",
      "uri": "https://...",
    },
    "imageAspectRatio": "square",
    "imageBackgroundColor": "#aaaaaa",
    "imageSize": "contain",
    "text": "world",
    "thumbnailImageUrl": "https://...",
    "title": "HELLO",
    "type": "buttons",
  },
  "type": "template",
}
`;

exports[`CarouselTemplate render match snapshot 1`] = `
Object {
  "altText": "xxx",
  "template": Object {
    "columns": Array [
      Object {
        "actions": Array [
          Object {
            "label": "with fries",
            "type": "uri",
            "uri": "https://...",
          },
          Object {
            "label": "with salad",
            "type": "uri",
            "uri": "https://...",
          },
        ],
        "defaultAction": undefined,
        "imageBackgroundColor": undefined,
        "text": "Burger",
        "thumbnailImageUrl": undefined,
        "title": undefined,
      },
      Object {
        "actions": Array [
          Object {
            "label": "with soup",
            "type": "uri",
            "uri": "https://...",
          },
          Object {
            "label": "with salad",
            "type": "uri",
            "uri": "https://...",
          },
        ],
        "defaultAction": undefined,
        "imageBackgroundColor": "#bbbbbb",
        "text": "Naporitan",
        "thumbnailImageUrl": "https://...",
        "title": "Pasta",
      },
    ],
    "imageAspectRatio": "square",
    "imageSize": "contain",
    "type": "carousel",
  },
  "type": "template",
}
`;

exports[`ConfirmTemplate render match snapshot 1`] = `
Object {
  "altText": "xxx",
  "template": Object {
    "actions": Array [
      Object {
        "label": "Blue pill",
        "type": "uri",
        "uri": "https://matrix.io/login",
      },
      Object {
        "label": "Red pill",
        "type": "uri",
        "uri": "https://matrix.io/leave",
      },
    ],
    "text": "Take a pill",
    "type": "confirm",
  },
  "type": "template",
}
`;

exports[`ImageCarouselTemplate render match snapshot 1`] = `
Object {
  "altText": "xxx",
  "template": Object {
    "columns": Array [
      Object {
        "action": Object {
          "label": "foo",
          "type": "uri",
          "uri": "https://...",
        },
        "imageUrl": "https://...",
      },
      Object {
        "action": Object {
          "label": "bar",
          "type": "uri",
          "uri": "https://...",
        },
        "imageUrl": "https://...",
      },
    ],
    "type": "image_carousel",
  },
  "type": "template",
}
`;
